src/client/dfuse/ops/readdir.c

```diff
@@ -24,240 +24,137 @@
+	size_t		b_offset;
+};
+
+int
+rd_cb(dfs_t *dfs, dfs_obj_t *dir, const char name[], void *_udata)
+{
+	struct rd_data	*udata = (struct rd_data *)_udata;
+	struct dfuse_projection_info *fs_handle = fuse_req_userdata(udata->req);
+	dfs_obj_t	*obj;
+	daos_obj_id_t	oid;
+	struct stat	stbuf = {0};
+	int		ns;
+	int		rc;
+
+	/*
+	 * MSC - from fuse fuse_add_direntry: "From the 'stbuf' argument the
+	 * st_ino field and bits 12-15 of the st_mode field are used. The other
+	 * fields are ignored." So we only need to lookup the entry for the
+	 * mode.
+	 */
+	DFUSE_TRA_DEBUG(udata->inode, "Adding entry name '%s'", name);
+	rc = dfs_lookup_rel(dfs, dir, name, O_RDONLY, &obj, &stbuf.st_mode);
+	if (rc)
+		return rc;
+
+	rc = dfs_obj2id(obj, &oid);
+	if (rc)
+		D_GOTO(out, rc = -rc);
+
+	rc = dfuse_lookup_inode(fs_handle, udata->inode->ie_dfs, &oid,
+				&stbuf.st_ino);
+	if (rc)
+		D_GOTO(out, rc = EIO);
+
+	ns = fuse_add_direntry(udata->req, udata->buf + udata->b_offset,
+			       udata->size - udata->b_offset, name, &stbuf,
+			       (off_t)(udata->anchor));
+
+	DFUSE_TRA_DEBUG(udata->inode, "add direntry: size = %ld, return %d\n",
+			udata->size - udata->b_offset, ns);
+
+	/*
+	 * This should be true since we accounted for the fuse_dirent size when
+	 * we started dfs_readdir_size().
+	 */
+	D_ASSERT(ns <= udata->size - udata->b_offset);
+	udata->b_offset += ns;
+
+out:
+	dfs_release(obj);
+	return rc;
+}
+
 void
 dfuse_cb_readdir(fuse_req_t req, struct dfuse_inode_entry *inode,
 		 size_t size, off_t offset)
 {
-	struct dfuse_projection_info	*fs_handle = fuse_req_userdata(req);
-	size_t				b_offset = 0;
-	daos_anchor_t			anchor = {0};
-	uint32_t			nr = LOOP_COUNT;
-	struct dirent			dirents[LOOP_COUNT];
-	int				next_offset = 0;
-	int				i;
-	void				*buf = NULL;
-	int				ns;
-	int				rc;
+	daos_anchor_t	*anchor = NULL;
+	uint32_t	nr = LOOP_COUNT;
+	void		*buf = NULL;
+	size_t		readdir_size, loop_size;
+	struct rd_data	*udata = NULL;
+	int		i;
+	int		rc;
 
-	DFUSE_TRA_DEBUG(inode, "Offset %zi",
-			offset);
+	DFUSE_TRA_DEBUG(inode, "Offset %zi", offset);
 
-	/* TODO:
-	 * To do this properly we need a way to convert from a offset into
-	 * a daos_anchor_t, in all cases.
-	 *
-	 * For now simply consume the first "offset" entries in a directory and
-	 * start iterating from there.  This will be correct only if the
-	 * directory contents aren't modified between calls.
-	 */
 	if (offset != 0) {
-		uint32_t count = nr;
-
-		DFUSE_TRA_ERROR(inode,
-				"Unable to correctly handle non-zero offsets");
-
-		while (offset > 0) {
-
-			if (offset < count) {
-				count = offset;
-			}
-
-			rc = dfs_readdir(inode->ie_dfs->dffs_dfs, inode->ie_obj,
-					 &anchor, &count, dirents);
-			if (rc != -DER_SUCCESS) {
-				D_GOTO(err, 0);
-			}
-			offset -= count;
-			next_offset += count;
+		anchor = (daos_anchor_t *)offset;
+		if (daos_anchor_is_eof(anchor)) {
+			D_FREE(anchor);
+			fuse_reply_buf(req, NULL, 0);
+			return;
 		}
+	} else {
+		D_ALLOC_PTR(anchor);
```

Comment:
It is not clear to me where this is freed. Could you please advise?

Reply:
5 lines above.
Basically i made the offset returned by fuse readdir to be the pointer to the anchor. fuse will keep calling readdir with that anchor until we hit EOF and return offset 0.
if Fuse stops calling readdir before we hit EOF, then we leak it, but i don't think that will be the case unless the fuse daemon crashes?

Reply:
yes i think it would leak.. i did some tests to check this.
i was thinking fuse might be doing itself preallocated buffers (4k) that it just calls regardless of the user's buffer. but it doesn't seem to be the case.

Reply:
hmm, the readdir callback function is missing 1 parameter (struct fuse_file_info *fi).
i don't know why Ashley didn't add that when the CB is issued. I think if we have that, we can store the anchor on the open handle of the fuse dir and free it when the dir is closed. I'll re-work this patch.